### PR TITLE
*: fix bugs when window function meets prepare

### LINKED ISSUE #9748: Unexpected error when window function meets prepare
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.

```
CREATE TABLE `t1` (
  `id` int(11) DEFAULT NULL,
  `sex` char(1) DEFAULT NULL
)
PREPARE p FROM "SELECT sex, COUNT(id) OVER (ORDER BY id ROWS BETWEEN ? PRECEDING and ? PRECEDING) FROM t1";
```

2. What did you expect to see?

prepare successfully.

3. What did you see instead?

error returned
```
mysql> PREPARE p FROM "SELECT sex, COUNT(id) OVER (ORDER BY id ROWS BETWEEN ? PRECEDING and ? PRECEDING) FROM t1";
ERROR 3586 (HY000): Window '<unnamed window>': frame start or end is negative, NULL or of non-integral type
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

master 06183393ed536aa248248b72d4bf9f5010947f66

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve? 

There're several problems when window function meets problem:
1. The `Name` of `ast.WindowSpec` is changed. It will cause error when execute a statement. fix 
1. We fill the paramMarker to NULL in when prepare it. It will cause bug in window function's frame bound.
fix #9748 
### What is changed and how it works?

Add a field to `paramMarker` so that we can know whether it's in prepare or in execute.
Then use the field to avoid some check if it's in prepare currently.

Don't change the `Name` field of `WindowSpec`, change it when necessary.

### Check List 

Tests 

 - Unit test (will add them later)

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

### CODE REVIEW

File: sessionctx/variable/tidb_vars.go [hunk 1]
```diff
@@ -289,7 +289,7 @@ const (
 	DefTiDBHashAggFinalConcurrency   = 4
 	DefTiDBForcePriority             = mysql.NoPriority
 	DefTiDBUseRadixJoin              = false
-	DefEnableWindowFunction          = false
+	DefEnableWindowFunction          = true
```
Comment (winoros) [inline sessionctx/variable/tidb_vars.go]: Oh it's a unnecessary change, will change back in next commit.

File: executor/prepared.go [hunk 2]
```diff
@@ -222,6 +226,7 @@ func (e *ExecuteExec) Build() error {
 	b := newExecutorBuilder(e.ctx, e.is)
 	stmtExec := b.build(e.plan)
 	if b.err != nil {
+		log.Warn("rebuild plan in EXECUTE statement failed", zap.String("name label of PREPARE statement", e.name))
```
Comment (lamxTyler) [inline executor/prepared.go]: `LabelName` is enough?

**Comment by winoros** [comment]
Adding now

File: planner/core/prepare_test.go [hunk 3]
```diff
@@ -318,3 +318,27 @@ func readGaugeInt(g prometheus.Gauge) int {
...
+		tk.MustExec("set @@tidb_enable_window_function = 0")
+	}()
+	tk.MustExec("use test")
+	tk.MustExec("create table window_prepare(a int, b double)")
+	tk.MustExec("insert into window_prepare values(1, 1.1), (2, 1.9)")
+	tk.MustExec("prepare stmt1 from 'select row_number() over() from window_prepare';")
+	// Test the unnamed window can be executed successfully.
+	tk.MustQuery("execute stmt1").Check(testkit.Rows("1", "2"))
+	// Test the stmt can be prepared successfully.
+	tk.MustExec("prepare stmt2 from 'select count(a) over (order by a rows between ? preceding and ? preceding) from window_prepare'")
```
Comment (eurekaka) [inline planner/core/prepare_test.go]: Add `execute stmt2 using xxx` test line also?

Comment (winoros) [reply planner/core/prepare_test.go]: added

**Pull request merged**